components/tidb_query_executors/src/runner.rs

```diff
@@ -424,10 +434,18 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
                 chunks.push(chunk);
             }
 
-            if drained {
+            if drained
+                || self
+                    .paging_size
+                    .map_or(false, |p| chunks.len() >= p as usize)
```

Comment:
I only changed this line from
`.map_or(false, |p| record_len >= p as usize)`
to
`.map_or(false, |p| chunks.len() >= p as usize)`

Because the stop condition does not make sense before though the performance looks great(stop when a batch scanned more rows than a batch? that's weird).

Then the duration of selection request increases, I tried changing it back, everything works fine :(

Reply:
I see. A `chunk` is a batch of data and `record_len` is how many rows are in this `chunk`. You should aggregate all chunks' length here.

Reply:
After adding some logs, I found the root cause. The duration of selection increases because of too large page. I was thought there will be more rows if using `record_len >= p`, that's wrong. Even if the client asked for a page with over 10,000 rows, 10 rows may be returned in this case.
If we use `chunks.len() >= p`, then for the first page, the default value will be 64, which would be a slow scan.
After I tuning the start paging size to 2, it's fine now. Anyway, that's out of this PR.